### PR TITLE
#14690: Reorganize gtests under `tests/tt_metal/tt_metal`

### PR DESCRIPTION
**Pull request by sagarwalTT**
### Ticket
#14690 

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/tt_metal/tt_metal/unit_tests/compute/test_broadcast.cpp
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/unit_tests/compute/test_broadcast.cpp:292]: currently everything in `unit_tests/compute` are meant to be simple single core LLK test. I think we should have a dedicated LLK or Compute fixture for all the tests in this directory that currently run using `DeviceFixture`

File: tests/tt_metal/tt_metal/unit_tests/dram/direct.cpp
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/unit_tests/dram/direct.cpp:375]: this should be moved to the bring-up folder

File: tests/tt_metal/tt_metal/unit_tests_common/basic/test_kernel_creation.cpp
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/unit_tests_common/basic/test_kernel_creation.cpp:5]: this should be moved to api constructs dir

**Review by abhullar-tt — COMMENTED** [review]
Looks good so far!

From going through this a couple categories that could help define the structure stick out to me:
- a bring-up / basic section in the unit_test dir and unit_test_common. These tests would go through minimal Metal RT codepath but what we need working before we try running more complex codepaths 
- LLK section, LLK team has been adding parameterized single core slow dispatch tests under "compute" 
- API constructs section: tests that create/initialize rt args, semaphores, circular buffers, kernels, programs, allocator (these wouldn't necessarily run anything on device but serve as the base complexity level of RT / flush out API rules)
- Fast dispatch tests: single chip/multi-chip enqueue APIs 
- Launching programs: these can be incrementally complex as Paul mentioned (test loading a kernel, tests that test across eth/tensix, complex use cases)

**Review by tt-rkim — APPROVED** [review]

**Review by aliuTT — APPROVED** [review]
Don't forget to squash all the commits

File: tests/tt_metal/tt_metal/common/command_queue_fixture.hpp
Comment (afuller-TT) [inline tests/tt_metal/tt_metal/common/command_queue_fixture.hpp:22]: Generally c'tor / d'tor are preferred over SetUp / TearDown.  https://google.github.io/googletest/faq.html#CtorVsSetUp

Is it required here because `CloseDevice` isn't `noexcept` (why isn't it `noexcept`? 🤔 )

File: CONTRIBUTING.md
Comment (afuller-TT) [inline CONTRIBUTING.md:257]: Looks like this is missing the filename of the binary.

**Review by afuller-TT — APPROVED** [review]
Nice!

File: tests/tt_metal/tt_metal/dispatch/dispatch_buffer/CMakeLists.txt
Comment (sagarwalTT) [reply tests/tt_metal/tt_metal/dispatch/dispatch_buffer/CMakeLists.txt]: I initially made this a cache variable because I wanted it to be visible to the `CMakeLists.txt` file in the parent directory, and the `CACHE` keyword allows variables to be globally accessible, however, since it only needs to be visible to the parent, I've just changed it to use `PARENT_SCOPE` instead.

Comment (afuller-TT) [reply tests/tt_metal/tt_metal/dispatch/dispatch_buffer/CMakeLists.txt]: Follow-up question: why does it need to be visible in the parent dir?  🙃 
In CMake generally "objects" should be encapsulated and know about themselves and connect to others like Lego pieces, rather than flinging variables around.

It looks like we're building 2 executables for each test?  One as a topic-focused, and then another "umbrella" executable that combines a bunch of topics?  Why do we want two binaries for the same test?

Currently that means each test file will be compiled twice.  Instead can create an `OBJECT` lib for the test files; they'll be compiled once.  Then define both executables listing the `OBJECT` targets as dependencies.  No duplicate compile, still two binaries, and bonus: no hucking variables to wider scopes.

**Comment by blozano-tt** [comment]
I am not going to block this PR, because I applaud that some reorganization work is being done here.

My only comment is: Since this is being reorganized ... why not put everything under `tt-metal/tt_metal/`. The tests are very clearly associated with the tt_metal library, why do they need to be outside of that directory in a global tests/ directory?

This is a minor gripe that @afuller-TT and I have.

If they are in the same tree it simplifies things like assigning code owners to tests.

**Review by blozano-tt — APPROVED** [review]

**Comment by ttmchiou** [comment]
is this PR ready for merge?
I checked and don't see any passing post-commit runs.

With these changes, I would be extra careful and run all the workflows to ensure all tests pass.

**Comment by sagarwalTT** [comment]
> is this PR ready for merge? I checked and don't see any passing post-commit runs.
> 
> With these changes, I would be extra careful and run all the workflows to ensure all tests pass.

I'm just fixing up some minor issues regarding failing pipeline runs. Once I fix those, this will be ready to merge.

This PR should only affect the all post-commit tests, TG unit tests, TGG unit tests and T3000 unit tests pipelines.

**Comment by tt-rkim** [comment]
Thanks boss, please post passing links once done

**Comment by sagarwalTT** [comment]
All post-commit pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/11991509497
T3000 pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/11991511278
TG pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/11991513649

**Review by rfurko-tt — APPROVED** [review]

**Pull request merged**
